src/chain/chaininterface.rs

```diff
@@ -54,44 +60,79 @@ pub enum ConfirmationTarget {
 /// called from inside the library in response to ChainListener events, P2P events, or timer
 /// events).
 pub trait FeeEstimator: Sync + Send {
-	fn get_est_sat_per_vbyte(&self, ConfirmationTarget) -> u64;
+	fn get_est_sat_per_vbyte(&self, confirmation_target: ConfirmationTarget) -> u64;
 }
 
 /// Utility to capture some common parts of ChainWatchInterface implementors.
 /// Keeping a local copy of this in a ChainWatchInterface implementor is likely useful.
 pub struct ChainWatchInterfaceUtil {
 	watched: Mutex<(Vec<Script>, Vec<(Sha256dHash, u32)>, bool)>, //TODO: Something clever to optimize this
 	listeners: Mutex<Vec<Weak<ChainListener>>>,
+	reentered: AtomicUsize
 }
 
-impl ChainWatchInterfaceUtil {
-	pub fn new() -> ChainWatchInterfaceUtil {
-		ChainWatchInterfaceUtil {
-			watched: Mutex::new((Vec::new(), Vec::new(), false)),
-			listeners: Mutex::new(Vec::new()),
-		}
-	}
-
-	pub fn install_watch_script(&self, spk: Script) {
+/// Register listener
+impl ChainWatchInterface for ChainWatchInterfaceUtil {
+	fn install_watch_script(&self, script_pub_key: Script) {
 		let mut watched = self.watched.lock().unwrap();
-		watched.0.push(Script::from(spk));
+		watched.0.push(Script::from(script_pub_key));
+		self.reentered.fetch_add(1, Ordering::Relaxed);
 	}
 
-	pub fn install_watch_outpoint(&self, outpoint: (Sha256dHash, u32)) {
+	fn install_watch_outpoint(&self, outpoint: (Sha256dHash, u32)) {
 		let mut watched = self.watched.lock().unwrap();
 		watched.1.push(outpoint);
+		self.reentered.fetch_add(1, Ordering::Relaxed);
 	}
 
-	pub fn watch_all_txn(&self) { //TODO: refcnt this?
+	fn watch_all_txn(&self) {
 		let mut watched = self.watched.lock().unwrap();
 		watched.2 = true;
+		self.reentered.fetch_add(1, Ordering::Relaxed);
 	}
 
-	pub fn register_listener(&self, listener: Weak<ChainListener>) {
+	fn register_listener(&self, listener: Weak<ChainListener>) {
 		let mut vec = self.listeners.lock().unwrap();
 		vec.push(listener);
 	}
+}
+
+impl ChainWatchInterfaceUtil {
+	pub fn new() -> ChainWatchInterfaceUtil {
+		ChainWatchInterfaceUtil {
+			watched: Mutex::new((Vec::new(), Vec::new(), false)),
+			listeners: Mutex::new(Vec::new()),
+			reentered: AtomicUsize::new(1)
+		}
+	}
+
+	/// notify listener that a block was connected
+	/// notification will repeat if notified listener register new listeners
+	pub fn block_connected(&self, block: &Block, height: u32) {
+		let mut watch = self.reentered.load(Ordering::Relaxed);
```

Comment:
Is relaxed actually OK here? If you have multiple threads calling fetch_add(1) are you allowed to see reentered go backwards? I assume not, but I dont think the performance hit of something stricter matters?

Reply:
enters are not strictly counted, the requirement is only to notice that there was a re-entry.